content/blog/2019-09-25-function-calls.md

```diff
@@ -0,0 +1,243 @@
+
+An `Argument` JSON object contains the argument's name and type:
+
+```typescript
+{"name": "<string>", "type": <Type>}
+```
+
+The JSON Bril `Program` object remains unchanged as a list of functions.
+
+### Compiling to JSON
+
+We extended the frontend for text-based Bril in `briltxt.py`.
+The goal was to convert our new function definitions and call instructions to the JSON representation of Bril, necessitating extending the parser and JSON generators.
+
+We also extended the TypeScript frontend in `ts2bril.ts`.
+The TypeScript parser already handled calls to handle treating `console.log` statements as Bril's `print` statements. We extended this component to also capture effectful calls that return results. In addition, the initial implementation did not support function clarations, so we added new support to transform the declaration and type information.
+
+### Interpreter
+
+The interpreter needed to be able to handle functions and calls in their extended JSON representation. 
+The main work was when encountering a `call` instruction: we create a new, empty environment with the arguments bound to the correct values.
+The interpreter searches for the function name in the program's list of functions since we are not implementing first-class functions.
+Because we chose to represent the stack implicitly, function calls are executed with a recursive call to `evalInstr`, thus relying on the underlying TypeScript stack frame implementation.
+
+Helpful compilers also need to check for errors. The interpreter now checks for a number of possible errors when calling functions. We use simple dynamic type checking to ensure that (1) argument types match the types of the provided values and (2) the function's declared return type matches both the type of the returned value and the type of the variable where the returned value is being stored.
+
+### Design decisions
+
+There were surprisingly many decisions to be made in the course of designing function calls.
+
+- For the sake of a sufficiently-scoped project, we chose not to implement first-order functions.
+- We implicitly represent the stack with recursive interpreter calls for simplicity based on the functionality we target.
+An explicit stack would allow more interesting control flow in the future.
+- We chose to allow backwards compatibility with the original Bril `main` syntax that did not have a return type or arguments.
+Similarly, the typescript `main` function is not explicitly demarcated&mdash; it is understood to consist of the instructions before any function definitions.
+- Calls can be effectful or non-effectful.
+In the JSON representation of Bril, we chose to represent `call` as its own 'kind' of instruction, allowing us to include the function's name as an explicit `name` field in the JSON object rather than an argument to the instruction.
```

Comment:
```suggestion
In the JSON representation of Bril, we chose to represent `call` as its own "kind" of instruction, allowing us to include the function's name as an explicit `name` field in the JSON object rather than an argument to the instruction.
```

Reply:
This is the trickiest decision you had to make, IMO—whether to "overload" the instruction arguments to include both the function name and the argument values.